### PR TITLE
New app task

### PR DESCRIPTION
**Pull request by davrodpin**
Add command to create empty grommet project

This change adds a new task to the Grommet command line tool that
creates a directory layout, html, js and scss files and all dependencies required to start writing a Grommet application.
The main difference between 'init' and 'new' commands is that a project created with 'new' will not contain any component other than a \<Main\> that contains an \<App\>.

Other minor changes:

- Remove gulp signal function (done function) that enables gulp task dependency since there is no dependency to be managed
- Using absolute paths so 'process.chdir' call could be removed

Depends-on: #667

### CODE REVIEW

File: bin/grommet.js [hunk 1]
```diff
@@ -17,76 +18,159 @@ String.prototype.capitalize = function() {
...
-          appName: app,
-          appTitle: title,
-          grommetVersion: packageJSON.version
-        }))
-        .pipe(gulp.dest('./'))
-        .pipe(install()).on('finish', function() {
-          done();
-        });
+      if (fileExists(newAppPath)) {
+        console.error('[grommet] Error while creating new app. Directory "'+command.options.app+'" already exists.');
```
Comment (alansouzati) [inline bin/grommet.js]: Can we add that in the init command too :)

Comment (alansouzati) [inline bin/grommet.js]: the spacing here as well: ' + command.options.app + '

File: bin/grommet.js [hunk 2]
```diff
@@ -17,76 +18,159 @@ String.prototype.capitalize = function() {
...
+        if (err) {
+          console.error('[grommet] Error trying to create project: ' + err);
+          process.exit(1);
+        }
+
+        try {
+          var templateFolder = path.join(command.env.templatePath, 'new/**');
+          copyAssetsAndInstall(templateFolder, newAppPath, command);
+        } catch(err) {
+          shelljs.rm('-rf', newAppPath);
```
Comment (alansouzati) [inline bin/grommet.js]: i like that we delete the project folder if anything fails. can we add that also inside the init command?

File: bin/grommet.js [hunk 3]
```diff
@@ -17,76 +18,159 @@ String.prototype.capitalize = function() {
...
+          console.error('[grommet] Error while creating new app '+command.options.app+'.');
```
Comment (alansouzati) [inline bin/grommet.js]: spacing here: ' + command.options.app + '

File: bin/grommet.js [hunk 4]
```diff
@@ -17,76 +18,159 @@ String.prototype.capitalize = function() {
...
+    command.options['title'] = command.options.app.replace(/-|_/g, ' ').capitalize();
+    break;
+  case 'new':
+    command.options['app'] = options[1];
+    if (!command.options.app) {
+      console.error('[grommet] Usage: grommet new <app-name> [description]');
+      process.exit(1);
+    }
+    command.options['title'] = command.options.app.replace(/-|_/g, ' ').capitalize();
+    command.options['description'] = options[2] || "";
```
Comment (alansouzati) [inline bin/grommet.js]: how you deal with spaces in the description ?

File: bin/grommet.js [hunk 5]
```diff
@@ -17,76 +18,159 @@ String.prototype.capitalize = function() {
...
-var command = options[0] || 'init';
-if (command === 'init') {
-  var app = options[1] || 'app-name';
-  var title = app.replace(/-|_/g, ' ').capitalize();
+switch(command.task) {
+  case 'init':
+    command.options['app'] = options[1] || 'app-name';
+    command.options['title'] = command.options.app.replace(/-|_/g, ' ').capitalize();
+    break;
+  case 'new':
```
Comment (alansouzati) [inline bin/grommet.js]: I have a use case that is coming up soon that is to generate a new project to be published in github (static site).

I think we could deprecate `init` and add one more attribute to new so that we could do:

`grommet new empty sample-app "Some description"`
`grommet new full sample-app "Some description"`
`grommet new static sample-app "Some description"`

File: templates/new/package.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "<%= appName %>",
+  "version": "0.1.0",
+  "main": "src/js/index.js",
```
Comment (alansouzati) [inline templates/new/package.json]: can we add the description field here as well? now that it is an option in the cli?

File: bin/grommet.js (hunk 4 above)
Comment (davrodpin) [inline bin/grommet.js]: > how you deal with spaces in the description ?

If the user wants to input a description that contains white spaces, he should surround the description with quotes, since the description is a command line option,

File: bin/grommet.js (hunk 5 above)
Comment (davrodpin) [inline bin/grommet.js]: > I have a use case that is coming up soon that is to generate a new project to be published in github (static site).
> 
> I think we could deprecate init and add one more attribute to new so that we could do:
> 
> grommet new empty sample-app "Some description"
> grommet new full sample-app "Some description"
> grommet new static sample-app "Some description"

Are *empty*, *full* and *static* some sort of template? If yes, I think it should be designed as an optional argument, as follows:

```
$ grommet new [-t <template-name>] <app-name> [<description>] 
```

In any case, don't you think this new change should be part of another pull request?

**Comment by davrodpin** [comment]
@alansouzati, PTAL

**Comment by alansouzati** [comment]
I would definitely be ok with that being resolved in another PR. my only concerned is the backward compatibility, but I guess it is ok since the the default task for `new` would make sense to be the empty project itself.

**Comment by davrodpin** [comment]
> I would definitely be ok with that being resolved in another PR. my only concern is the backward compatibility, but I guess it is ok since the the default task for new would make sense to be the empty project itself.

Ow, I see your point, but I don't think there will break on the backward compatibility:

Command to create a new empty app:

```
$ grommet new my-sample-app "Sample App"
```

After add support for templates, we would add an optional argument like this:

```
grommet new my-sample-app -t empty "Sample App"
```

Note that *-t* is optional and without specifying it, the grommet command line tool would use the *empty* template as default.

**Comment by alansouzati** [comment]
correct, makes sense :)

**Comment by alansouzati** [comment]
Thanks for your contribution
